src/databricks/sql/client.py

```diff
@@ -297,6 +298,58 @@ def _check_not_closed(self):
         if not self.open:
             raise Error("Attempting operation on closed cursor")
 
+    def _handle_staging_operation(self):
+        """Make HTTP request using instructions provided by server"""
+
+        row = self.active_result_set.fetchone()
+
+        # TODO: Handle headers. What format will gateway send? json? plaintext?
+        operation, presigned_url, local_file, headers = (
+            row.operation,
+            row.presignedUrl,
+            row.localFile,
+            None,
+        )
+
+        operation_map = {
+            "PUT": requests.put,
+            "GET": requests.get,
+        }
+
+        if operation not in operation_map:
+            raise Error(
+                "Operation {} is not supported. Supported operations are {}".format(
+                    operation, ",".join(operation_map.keys())
+                )
+            )
+
+        req_func = operation_map[operation]
+
+        if local_file:
+            raw_data = open(local_file, "rb")
+        else:
+            raw_data = None
+
+        rq_func_args = dict(url=presigned_url, data=raw_data)
+
+        logger.debug(
+            "Attempting staging operation: {} - {}".format(operation, local_file)
+        )
+
+        # Call the function
+        resp = req_func(**rq_func_args)
+
+        if resp.status_code != 200:
+            raise Error(
+                "Staging operation over HTTP was unsuccessful: {}-{}".format(
```

Comment:
there are different failure scenarios:
1. non retriable errors (e.g., 401 or 403, file system access error)
2. retryable errors (e.g., socket connection failure)

we should consider retrying retryable failures. how do we retry idempotent query failures? what is our retry policy for query failures?

Reply:
Good questions. I don't think this should block merge here though. We can add it with a follow-up PR subsequent.

PySQL has robust retry behaviour specified for **thrift operations**, including unit and integration tests. But these staging operations are not performed over Thrift. So if we implement retry behaviour here we'll be doing so from scratch and will should probably send it through a one-decision doc + review because each cloud provider may have different needs / response headers etc. 

For now, the http request to the cloud storage provider is tried exactly once. In a follow-up we can implement retry behaviour for the ingestion HTTP Requests (and probably even borrow the config used for thrift retries)

Reply:
The error is more precise if we specify it's the pre-signed url over HTTP and we should anticipate this can fail.
However, for retryable errors, I am wondering how much the client lib here should retry.

it's possible that the user of this client lib will retry as well, we should be consistent with other places to handle the transit error.

FYI,
PUT url is createOrUpdate, so it's ok to retry.
Delete url can be issued multiple times and only take effect when file is found.

Reply:
I consider this outside of the scope of this PR. we should follow up on retry behaviour later after this PR.